intake_netflow/v9.py

```diff
@@ -0,0 +1,521 @@
+        self.flowsets = flowsets
+
+    def update_cache(self, cache):
+        """Update cache of template records."""
+        for i, flowset in enumerate(self.flowsets):
+            if not isinstance(flowset, TemplateFlowSet):
+                continue
+            for id, record in flowset.templates.items():
+                cache[id] = record
+
+    def apply(self, templates):
+        """Deserialize partially-decoded data flowsets.
+
+        Deserialization of a data flowset is a two-step process because we
+        cannot assume the needed template is available when we encounter the
+        data flowset. Thus, we place the deserialization process on hold into
+        a packet is fully read. Then we re-scan the partially-decoded data
+        flowsets and finish deserialization.
+        """
+        for i, flowset in enumerate(self.flowsets):
+            if isinstance(flowset, functools.partial):
+                self.flowsets[i] = flowset(templates)
+
+    @staticmethod
+    def decode(source):
+        header = Header.decode(source)
+        flowsets = [decode_flowset(source) for _ in range(header.count)]
+        return ExportPacket(flowsets, header=header)
+
+    def encode(self):
+        raw = self.header.encode()
+        for flowset in self.flowsets:
+            if isinstance(flowset, functools.partial):
+                continue
+            raw += flowset.encode()
+        return raw
+
+
+class PacketStream(object):
+    def __init__(self, source):
+        """A read-only representation of serialized packets.
+
+        Parameters:
+            source : io.BufferedReader
+                Read-only input for packets.
+        """
+        self._source = source
+        self._cache = {}
+
+    def next(self):
+        if self._source.peek() == b'':
+            raise StopIteration
+
+        # Packet deserialization is a two-step process.
+        #
+        # A data flowset requires a defined template to complete the
+        # deserialization process, Since template and data flowsets can be
+        # out-of-order within a packet, then we must temporarily wait to
+        # finish deserializing a data flowset and the respective records.
+
+        packet = ExportPacket.decode(self._source)
+
+        # Add templates to cache
+        packet.update_cache(self._cache)
+
+        # Finish deserialization
+        packet.apply(self._cache)
+
+        return packet
+
+    def __next__(self):
+        return self.next()
+
+    def __iter__(self):
+        return self
+
+    def close(self):
+        return self._source.close()
+
+
+class RecordStream(PacketStream):
+    def __init__(self, source):
+        """A read-only representation of serialized data records.
+
+        Parameters:
+            source : io.BufferedReader
+                Read-only input for data records.
+        """
+        super(RecordStream, self).__init__(source)
+        self._queue = []
+
+    def next(self):
+        while len(self._queue) == 0:
```

Comment:
The parent class, `PacketStream`, will throw a StopIteration exception when the underlying stream EOFs. Otherwise, `RecordStream` will find a packet with the needed data.

The reason for this logic is that the Netflow spec allows packets to contain only templates, which we are ignoring in this instance. We need to read the packet stream until we find a data flowset containing the actual records.

Templates only make a small percentage of the overall traffic so the only case when the loop won't terminate is if we have an infinite stream of templates.